ydb/core/blobstorage/vdisk/hulldb/base/hullds_heap_it.h

```diff
@@ -0,0 +1,190 @@
+    protected:
+        struct THeapItem {
+            //using TLevelIndexSnapshotForwardIt = TLevelIndexSnapshot<TKey, TMemRec>::TForwardIterator*;
+            using TFreshIndexAndDataIterator = std::conditional_t<Forward, typename TFreshIndexAndDataSnapshot<TKey, TMemRec>::TForwardIterator,
+                                                                  typename TFreshIndexAndDataSnapshot<TKey, TMemRec>::TBackwardIterator>;
+            using TIterators = std::variant<TFreshIndexAndDataIterator*, typename TFreshAppendix<TKey, TMemRec>::TIterator*,
+                                            typename TOrderedLevelSegments<TKey, TMemRec>::TReadIterator*,
+                                            typename TLevelSegment<TKey, TMemRec>::TMemIterator*>;
+            TKey Key;
+            TIterators Iter;
+
+            friend bool operator<(const THeapItem& x, const THeapItem& y) {
+                if constexpr (Forward) {
+                    return y.Key < x.Key;
+                }
+                else{
+                    return x.Key < y.Key;
+                }
+            }
+
+            template <class TIter>
+            THeapItem(TIter *iter) {
+                Iter = iter;
+            }
+        };
+
+        std::vector<THeapItem> Heap;
+        size_t HeapItems = 0;
+
+        template<typename TMerger>
+        void Move(TMerger merger) {
+            Y_ABORT_UNLESS(Valid());
+            const TKey key = Heap.front().Key;
+            while (HeapItems && Heap.front().Key == key) {
+                std::pop_heap(Heap.begin(), Heap.begin() + HeapItems);
+                THeapItem& item = Heap[HeapItems - 1];
+                const bool remainsValid = std::visit([&](auto *iter) {
+                    if constexpr (!std::is_same_v<TMerger, std::nullptr_t>) {
+                        iter->PutToMerger(merger);
+                    }
+                    if constexpr (Forward) {
+                        iter->Next();
+                    }
+                    else{
+                        iter->Prev();
+                    }
+                    if (iter->Valid()) {
+                        item.Key = iter->GetCurKey();
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }, item.Iter);
+                if (remainsValid) {
+                    std::push_heap(Heap.begin(), Heap.begin() + HeapItems);
+                } else {
+                    --HeapItems;
+                }
+            }
+        }
+
+    public:
+        bool Valid() const {
+            return HeapItems;
+        }
+
+        template <class TIter>
+        void Add(TIter* iter) {
+            THeapItem item(iter);
+            Heap.push_back(item);
+            HeapItems++;
+        }
+
+        template<typename TMerger>
+        void PutToMergerAndAdvance(TMerger *merger) {
+            Move(merger);
+        }
+
+        void Next() {
```

Comment:
Next on backward iterator may cause some unexpected behaviour, same for Prev -- either check direct in runtime or in compile time (the second one is better)